### ISSUE TITLE
Build on armv7l fails

### ISSUE THREAD

**Issue by junghans** [STEllAR-GROUP/hpx#3509]
## Expected Behavior

Builds

## Actual Behavior

```
[  213s] [  0%] Building CXX object src/CMakeFiles/hpx.dir/exception.cpp.o
[  213s] cd /home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1/build/src && /usr/bin/c++  -DBOOST_DISABLE_ASSERTS -DHPX_COMPONENT_NAME=hpx -DHPX_DISABLE_ASSERTS -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_GNU_SOURCE -Dhpx_EXPORTS -I/home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1 -I/home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1/build  -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -DNDEBUG -O2 -g -DNDEBUG -fPIC   -pthread -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=trampolines -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Wno-unused-but-set-parameter -Wno-unused-but-set-variable -Wno-maybe-uninitialized -Wno-sync-nand -Wno-delete-non-virtual-dtor -fvisibility=hidden -std=c++17 -o CMakeFiles/hpx.dir/exception.cpp.o -c /home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1/src/exception.cpp
[  222s] In file included from /home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1/hpx/runtime.hpp:12,
[  222s]                  from /home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1/src/exception.cpp:14:
[  222s] /home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1/hpx/lcos/local/spinlock.hpp: In member function 'bool hpx::lcos::local::spinlock::acquire_lock()':
[  222s] /home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1/hpx/lcos/local/spinlock.hpp:112:31: error: 'BOOST_INTERLOCKED_EXCHANGE' was not declared in this scope
[  222s]              std::uint64_t r = BOOST_INTERLOCKED_EXCHANGE(&v_, 1);
[  222s]                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
[  222s] /home/abuild/rpmbuild/BUILD/hpx-1.2.0-rc1/hpx/lcos/local/spinlock.hpp:112:31: note: suggested alternative: 'BOOST_INTRUSIVE_DOCIGN'
[  222s]              std::uint64_t r = BOOST_INTERLOCKED_EXCHANGE(&v_, 1);
[  222s]                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
[  222s]                                BOOST_INTRUSIVE_DOCIGN
[  233s] make[2]: *** [src/CMakeFiles/hpx.dir/build.make:79: src/CMakeFiles/hpx.dir/exception.cpp.o] Error 1
```

## Steps to Reproduce the Problem

Try to build on armv7l, details [here](https://github.com/STEllAR-GROUP/hpx/files/2531927/build_arm7.txt).

## Specifications

  - HPX Version: 1.2.0-rc1
  - Platform (compiler, OS): gcc-8.2.1, OpenSuse

**Comment by diehlpk** [comment]
I added `HPX_WITH_GENERIC_CONTEXT_COROUTINES=ON` to the arm build and the error remains.

**Comment by hkaiser** [comment]
This is strange as the code in question will be compiled only if `BOOST_SP_HAS_SYNC` was not defined (see: ). However on ARM architecture this shouldn't be the case (see: https://github.com/boostorg/smart_ptr/blob/develop/include/boost/smart_ptr/detail/sp_has_sync.hpp#L25-L29). 

Could you please check whether the compiler you use defines `__GCC_HAVE_SYNC_COMPARE_AND_SWAP_4`?

**Comment by sithhell** [comment]
the problem is with us: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/lcos/local/spinlock.hpp#L31
We don't include the header if `__arm__` is defined.
This should be overhauled, I think.

**Comment by hkaiser** [comment]
@sithhell how did you build things on Android with this?

**Comment by sithhell** [comment]
I haven't built on android for a while. My other arm machines are aarch64, and the cross compiler does not define `__arm__` for that.

**Comment by junghans** [comment]
After #3557 and #3551, this is the last arch missing for Fedora and OpenSUSE, any news here?

**Comment by biddisco** [comment]
I have built recently on my Raspberry Pi 3. That's armv71 isn't it? Should I try getting a build going again and see if I can make it work?

**Comment by hkaiser** [comment]
> I have built recently on my Raspberry Pi 3. That's armv71 isn't it? Should I try getting a build going again and see if I can make it work?

That would make sense. However I'd suspect that your cross-compiler will not define `__arm__` either.

**Comment by junghans** [comment]
I just started a rebuild on OpenSUSE to check what the error is now using generic context, will report back.

**Comment by junghans** [comment]
Still the same:
```
[  263s] [  0%] Building CXX object src/CMakeFiles/hpx.dir/error_code.cpp.o
[  263s] cd /home/abuild/rpmbuild/BUILD/hpx_1.2.0/build/src && /usr/bin/c++  -DBOOST_DISABLE_ASSERTS -DHPX_COMPONENT_NAME=hpx -DHPX_DISABLE_ASSERTS -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_GNU_SOURCE -Dhpx_EXPORTS -I/home/abuild/rpmbuild/BUILD/hpx_1.2.0 -I/home/abuild/rpmbuild/BUILD/hpx_1.2.0/build  -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -DNDEBUG -O2 -g -DNDEBUG -fPIC   -pthread -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=trampolines -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Wno-unused-but-set-parameter -Wno-unused-but-set-variable -Wno-maybe-uninitialized -Wno-sync-nand -Wno-delete-non-virtual-dtor -fvisibility=hidden -std=c++17 -o CMakeFiles/hpx.dir/error_code.cpp.o -c /home/abuild/rpmbuild/BUILD/hpx_1.2.0/src/error_code.cpp
[  270s] [  0%] Building CXX object src/CMakeFiles/hpx.dir/exception.cpp.o
[  270s] cd /home/abuild/rpmbuild/BUILD/hpx_1.2.0/build/src && /usr/bin/c++  -DBOOST_DISABLE_ASSERTS -DHPX_COMPONENT_NAME=hpx -DHPX_DISABLE_ASSERTS -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_GNU_SOURCE -Dhpx_EXPORTS -I/home/abuild/rpmbuild/BUILD/hpx_1.2.0 -I/home/abuild/rpmbuild/BUILD/hpx_1.2.0/build  -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -DNDEBUG -O2 -g -DNDEBUG -fPIC   -pthread -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=trampolines -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Wno-unused-but-set-parameter -Wno-unused-but-set-variable -Wno-maybe-uninitialized -Wno-sync-nand -Wno-delete-non-virtual-dtor -fvisibility=hidden -std=c++17 -o CMakeFiles/hpx.dir/exception.cpp.o -c /home/abuild/rpmbuild/BUILD/hpx_1.2.0/src/exception.cpp
[  281s] In file included from /home/abuild/rpmbuild/BUILD/hpx_1.2.0/hpx/runtime.hpp:12,
[  281s]                  from /home/abuild/rpmbuild/BUILD/hpx_1.2.0/src/exception.cpp:14:
[  281s] /home/abuild/rpmbuild/BUILD/hpx_1.2.0/hpx/lcos/local/spinlock.hpp: In member function 'bool hpx::lcos::local::spinlock::acquire_lock()':
[  281s] /home/abuild/rpmbuild/BUILD/hpx_1.2.0/hpx/lcos/local/spinlock.hpp:112:31: error: 'BOOST_INTERLOCKED_EXCHANGE' was not declared in this scope
[  281s]              std::uint64_t r = BOOST_INTERLOCKED_EXCHANGE(&v_, 1);
[  281s]                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
[  281s] /home/abuild/rpmbuild/BUILD/hpx_1.2.0/hpx/lcos/local/spinlock.hpp:112:31: note: suggested alternative: 'BOOST_INTRUSIVE_DOCIGN'
[  281s]              std::uint64_t r = BOOST_INTERLOCKED_EXCHANGE(&v_, 1);
[  281s]                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
[  281s]                                BOOST_INTRUSIVE_DOCIGN
[  294s] make[2]: *** [src/CMakeFiles/hpx.dir/build.make:79: src/CMakeFiles/hpx.dir/exception.cpp.o] Error 1
[  294s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/hpx_1.2.0/build'
[  294s] make[1]: *** [CMakeFiles/Makefile2:1832: src/CMakeFiles/hpx.dir/all] Error 2
[  294s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/hpx_1.2.0/build'
[  294s] make: *** [Makefile:144: all] Error 2
[  294s] error: Bad exit status from /var/tmp/rpm-tmp.Fqz8FJ (%build)
```

**Comment by junghans** [comment]
To reproduce this issue, just branch https://build.opensuse.org/package/show/science:HPC/hpx and drop the following line from `hpx.spec`:
```spec
ExcludeArch: %arm
```

**Comment by diehlpk** [comment]
I am not wrong the raspberry pi 3 should be aarch64. And we know that it is building there from Fedora.

**Comment by junghans** [comment]
armv7l is 32-bit, while aarch64 is 64-bit. aarch64 worked, but `%arm` only refers to 32-bit.

**Comment by diehlpk** [comment]
@junghans I started to work on this, but compilation on my pi is so slow. Might take some time to get things fixed.

**Comment by sithhell** [comment]
This issue should appear when using a cross compilation toolchain as well.

Am Fr., 7. Dez. 2018 18:08 hat Patrick Diehl <notifications@github.com>
geschrieben:

**Comment by biddisco** [comment]
'bus error' on arm means that an unaligned access is occurring. I've had this before and fixed it a long time ago. If memory serves, we had one boost related problem and one hpx one, but for the life of me I cannot remember what they were. I will investigate this evening and see if I can find what the problem was.

**Comment by junghans** [comment]
`-march=native` is a no-go for rpm packages as every distro injects their own optimization flags!

**Comment by diehlpk** [comment]
@junghans Yes, I know, my priority for now is to get it compiled directly on the PI for a different project. Later, I will focus on the Fedora package.

**Comment by diehlpk** [comment]
@biddisco Could you please share the CMake Options you used to build HPX? Did you do a cross compilation or directly on the PI?

**Comment by diehlpk** [comment]
> 'bus error' on arm means that an unaligned access is occurring. I've had this before and fixed it a long time ago. If memory serves, we had one boost related problem and one hpx one, but for the life of me I cannot remember what they were. I will investigate this evening and see if I can find what the problem was.

@biddisco  https://github.com/boostorg/lockfree/issues/11#issuecomment-209777682

**Comment by diehlpk** [comment]
@junghans I will try to do a new build for the Fedora rpm package next week.

**Comment by diehlpk** [comment]
Once #3591 is merged, this issue can be closed.

**Comment by junghans** [comment]
@diehlpk, it would be nice to fix #3511 as well.

**Comment by diehlpk** [comment]
@junghans I would need to setup a cross compilation to work on this or I would need to access to one node to fix the compilation there. I could do it later next year, but right now it is not on my priority list.

**Comment by junghans** [comment]
@diehlpk, or just play around on openSUSE build, it was s390x VMs, too. Looking at https://github.com/STEllAR-GROUP/hpx/issues/3511#issuecomment-436376077 there is only a `__S390x` define missing somewhere in `util/hardware/timestamp.hpp`.

**Comment by diehlpk** [comment]
@junghans Ok, I maybe will have a look later next year.

**Comment by diehlpk** [comment]
@hkaiser This issues can be closed, due to #3591

**Comment by junghans** [comment]
Github's auto-magic will close it once #3591 got merged.
